ddtrace/contrib/tornado/middlewares.py

```diff
@@ -0,0 +1,67 @@
+from tornado.web import Application
+from tornado.stack_context import StackContext
+
+from . import handlers
+from .stack_context import ContextManager
+from ...ext import AppTypes
+
+
+class TraceMiddleware(object):
+    """
+    Tornado middleware class that wraps a Tornado ``HTTPServer`` instance
+    so that the request_callback can be wrapped with a ``StackContext``
+    that uses the internal ``ContextManager``. This middleware creates
+    a root span for each request.
+    """
+    def __init__(self, http_server, tracer, service='tornado-web'):
+        """
+        Replace the default ``HTTPServer`` request callback with this
+        class instance that is callable. If the given request callback
+        is a Tornado ``Application``, all handlers are wrapped with
+        tracing methods.
+        """
+        self._http_server = http_server
+        self._tracer = tracer
+        self._service = service
+        # the default http_server callback must be preserved
+        self._request_callback = http_server.request_callback
+
+        # the middleware instance is callable so it behaves
+        # like a regular request handler
+        http_server.request_callback = self
+
+        # configure the current service
+        self._tracer.set_service_info(
+            service=service,
+            app='tornado',
+            app_type=AppTypes.web,
+        )
+
+        if isinstance(self._request_callback, Application):
+            # request handler is a Tornado web app and we can safely wrap it
+            app = self._request_callback
+            for _, specs in app.handlers:
+                for spec in specs:
+                    self._wrap_application_handlers(spec.handler_class)
+
+    def _wrap_application_handlers(self, cls):
+        """
+        Wraps the Application class handler with tracing methods.
+        """
+        cls.on_finish = handlers.wrapper_on_finish(cls.on_finish)
```

Comment:
The other place where things would want a hook into is `Handler._handle_request_exception`. It's the only way I could see to get the actual exception information included in the trace. http://www.tornadoweb.org/en/stable/_modules/tornado/web.html & search for `def _handle_request_exception`. You could potentially use `def log_exception` if it gets called in all of the needed situations, haven't looked into that. At the very least it seems cleaner given that it's intended to be overridden. 

Anyway, the call into on_finish doesn't have enough information to actually grab the details of the exception for inclusion in the span. I was using a dedicated exception tracking system so I didn't get to the point of integrating thais w/APM, but had I not being using the other system I would have been looking for APM to record that information.